WIP Refactor tools

gzm0 opened the pull request.

gzm0 commented:
Review by @sjrd

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2166/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2166/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2801/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2801/</a>
Test FAILed.

gzm0 commented:
Jenkins, test this please

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2167/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2167/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2802/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2802/</a>
Test PASSed.

gzm0 commented:
I tested this locally.

sjrd commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Linker.scala:
```diff
@@ -38,50 +38,51 @@ final class Linker(semantics: Semantics, outputMode: OutputMode,
+  def clean(): Unit = ()
+
+  // Binary compat
+  def link(irInput: immutable.Traversable[VirtualScalaJSIRFile], logger: Logger,
```
Please add "TODO" in the comment, so that this can be found when we grep for "TODO"s and "deprecated"s one day.

sjrd commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Linker.scala:
```diff
@@ -38,50 +38,51 @@ final class Linker(semantics: Semantics, outputMode: OutputMode,
+    }
+
+    val infos = infosBuilder.result
+    val encodedNameToFile = encodedNameToFileBuilder.result
```
`result()` with `()` (twice). Calling `result()` has side effects, in theory, since it invalidates the state of the builder.

sjrd commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRFileCache.scala:
```diff
@@ -0,0 +1,155 @@
+  /** Returns the current set of files for use
+   *
+   *  Can only be invoked after an invocation of update (but an arbitrary
+   *  number of times).
```
and before `cleanAfterUse()`?

sjrd commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/IRFileCache.scala:
```diff
@@ -0,0 +1,155 @@
+      new IRFileCache.Stats(statsReused, statsInvalidated, statsTreesRead)
+
+    // "Garbage-collect" persisted versions of files that have disappeared
+    irCache.retain((_, f) => f.cleanAfterUse())
```
Can't we do this garbage collection in `update()`? That would avoid the `.filter(_.exists)` in `def files`.

sjrd replied:
And, btw, would completely eliminate the need for `cleanAfterUse()`. You could instead clear the stats at the beginning of `update`, and provide a side-effect-free accessor to the stats.

gzm0 replied:
Yes. We could do that. I first thought it restricts us from changing the caching strategy, but it doesn't seem to actually do so.
The only downside is that you cannot clean up the cache without updating it. So we might use more memory during the late phase of the pipeline (since we retain everything until the next run). Basically this means this patch potentially *increases* memory usage. Your call.

sjrd replied:
Er ... I meant to `retain` at the *end* of `update`. Not at the beginning.
```scala
val updated = irFiles.foldLeft(...)

irCache.retain(...)

updated
```
How would that increase memory usage, compared to doing exactly the same retain later in the algorithm, but in the same run? On the contrary, potentially this reduces memory retention earlier.

Unless I am completely missing something.

gzm0 replied:
Ah... Yes we could do this. Two points:

- This will not allow us to change the caching strategy (for example right now, we *never* release trees to GC, even if they are not used in a run).
- We would need to be smarter about when to release the original IR file to GC. The `PersistentIRFile` can definitely be smarter about it (don't keep a reference if you have everything you need). However, references to files whose tree is never read cannot seem to be released (which is the majority IIUC).

This is not an argument against moving the retain call into the `update` method, but an argument against removing the `cleanAfterUse` method. It is probably a good idea to put the `retain` call there, since the `filter` call in the `files` method likely creates a new strict collection (but who knows... Scala collection library FTW)...

sjrd replied:
OK let's keep `cleanAfterUse`, just in case. It's easier to deprecate it than to impose new calls.

sjrd commented on tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/ScalaJSOptimizer.scala:
```diff
@@ -81,29 +84,50 @@ class ScalaJSOptimizer(val semantics: Semantics, val outputMode: OutputMode,
+  def optimizeIR[T](buildJS: (JSTreeBuilder => Unit) => T)(
+      hasOldResult: () => Boolean)(
+      irFiles: Traversable[VirtualScalaJSIRFile],
+      cfg: OptimizerConfig, logger: Logger): Option[T] = {
```
Consider reordering the parameter lists as follows:
```scala
(irFiles, cfg, logger)(hasOldResult)(buildJS)
```
The rationale is 1. put lambdas at the end, as is more common in Scala APIs and 2. put `hasOldResult` first because it's called first.

gzm0 replied:
I can put `hasOldResult` earlier, but I cannot put `(irFiles, cfg, logger)` first, since it then would collide with another overload...

sjrd replied:
Erf... maybe that's a sign we should use a different name?

gzm0 replied:
Hahaha, touché. Suggestions? `optimizeIRInternal` comes to mind, but I dislike it, since it is not internal...

sjrd replied:
Good question. Back in my Delphi days the answer would be `OptimizeIREx` ^^
Maybe `optimizeIRIfChanged` or `optimizeIRCond`?

sjrd commented:
That's all.

gzm0 commented:
There is one thing I do not like about this solution at all btw.

Consider the following:

```scala
val opt = new ScalaJSOptimizer

val output1 = // ...
val output2 = // ...

def write(cp: IRClasspath) = {
  opt.optimizeIR(cp, output1)
  opt.optimizeIR(cp. output2)
}

write(cp1)
write(cp2) // updated classpath
```

The second time `write` is invoked, `output2` will not be written anymore. We could restrict the usage of an optimizer to a single output file, but there are two things I don't like about this:

1. The output file should then be an argument to the optimizer and not to the optimize method
2. In other cases where an option should be an optimizer argument but this is impractical to the user, we have added reset logic to the optimizer (n.b. `wantSourceMap`).

Maybe we should add a similar guard and only short circuit if the following three conditions are met:

1. The IR didn't change since the last run
2. The result file exists
3. (new) the result file is the same as last run

gzm0 commented:
Updated. Please don't merge yet, I'd like to have your opinion on my last comment first.

sjrd commented:
> Maybe we should add a similar guard and only short circuit if the following three conditions are met:

Yes that sounds good to me.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2174/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2174/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2813/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2813/</a>
Test PASSed.

gzm0 commented:
@sjrd I added an example commit with a guard. This yields to another problem: Now there is state in the optimizer which is not always updated depending on the method. So having a `ScalaJSClosureOptimizer` that uses a `ScalaJSOptimizer` that is actually used will now yield bad behavior:

```scala
val opt = new ScalaJSOptimizer
val fullOpt = new ScalaJSClosureOptimizer(opt)
opt.optimize()
// change in code
fullOpt.optimize()
opt.optimize() // boom, not updated
```

I'm starting to have the feeling we need to look at this from another angle... :S

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2176/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2176/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2819/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2819/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2175/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2175/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2818/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2818/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2179/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2179/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2826/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2826/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2180/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2180/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2828/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2828/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2181/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/2181/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2830/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/2830/</a>
Test FAILed.

gzm0 commented:
This is stale. I am working on this, but not on this branch and it will look very different. Closing.
